File: ez_supermarket/ez_supermarket/custom/item/item.js
// Copyright (c) 2023, Balamurugan and contributors
// For license information, please see license.txt

frappe.ui.form.on("Item", {
  setup: function (frm) {
    const warehouseFilters = {
      is_group: "0",
    };

    frm.set_query("custom_default_store_warehouse", function () {
      return { filters: warehouseFilters };
    });

    frm.set_query("custom_default_stall_warehouse", function () {
      return { filters: warehouseFilters };
    });

    frm.set_query("custom_default_store_location", function (doc) {
      return { filters: { warehouse: doc.custom_default_store_warehouse } };
    });

    frm.set_query("custom_default_stall_location", function (doc) {
      return { filters: { warehouse: doc.custom_default_stall_warehouse } };
    });
  },
  validate: function (frm) {
    console.log("Validation of Item Form");

    frappe.call({
      method:
        "ez_supermarket.ez_supermarket.custom.item.item.apply_tax_template",
      args: {
        doc_json: JSON.stringify(frm.doc),
      },
      callback: function (r) {
        console.log("Server response:", r);
        // Update the form with the response
        frm.doc = r.message;
      },
    });
  },
  custom_mrp: function (frm) {
    var value = frm.doc.custom_mrp;
    frm.set_value("standard_rate", value);
    frm.set_value("valuation_rate", value);
  },
});

File: ez_supermarket/ez_supermarket/custom/item/item.py
# Copyright (c) 2023, Balamurugan and contributors
# For license information, please see license.txt

import frappe
import json
from frappe import _
from frappe.utils import add_months, today
from datetime import datetime
from datetime import datetime, timedelta

# Server method
@frappe.whitelist()
def apply_tax_template(doc_json):
    doc = json.loads(doc_json)

    # Clear existing tax rows
    doc["taxes"] = []

    tax_rate = doc.get("custom_tax_rate")
    if tax_rate:
        # Fetch the company name from the 'item_defaults' table
        company = doc.get("item_defaults")[0].get("company") if doc.get("item_defaults") else None

        if company:
            # Fetch the company abbreviation
            company_abbr = frappe.db.get_value("Company", company, "abbr")

            if not company_abbr:
                frappe.throw(_("Could not find abbreviation for company: {0}").format(company))

                    # Create a new row in the taxes child table for In-State tax category
            tax_row_in_state = {
                "item_tax_template": "GST {}% - {}".format(tax_rate, company_abbr),
                "tax_category": "In-State"
            }
            doc["taxes"].append(tax_row_in_state)

            # Create another row in the taxes child table for Out-State tax category
            tax_row_out_state = {
                "item_tax_template": "GST {}% - {}".format(tax_rate, company_abbr),
                "tax_category": "Out-State"
            }
            doc["taxes"].append(tax_row_out_state)

    return doc